### PR TITLE
hevm: add knowledge to path conditions when query is Known

### LINKED ISSUE #456: non-derministic branching when backstepping during symbolic execution
I came across some bytecode that seems to have non deterministic branching behaviour. If I step forwards through the program I see only one branch point, but if I step forwards and then backwards and then forwards again across a `JUMPI` I am asked to choose a branch when stepping forwards across the `JUMPI` for the second time only.

I can only reproduce this behaviour after backstepping. It seems to be reproducible with `z3` and `cvc4`, although in the test program below I observed the non deterministic branching with `z3` at both PC13 and PC18, and with `cvc4` at PC13 only.

The test program was built from the following yul code:

```
{
    calldatacopy(0,0,1024)
    let y := mload(0x20)
    for {} and(y, 8) { pop(y) } {
        if y { continue }
        y := 0
    }
}
```

*Steps To Reproduce*

1. `hevm symbolic --code 61040060006000376020515b60088116156027578015601c576021565b600090505b8050600b565b50 --debug`
2. Step forwards until the first `JUMPI` at pc 13
3. Press `0` (do not jump)
4. Step forwards and backwards through the loop a few times

*Results*

After a while the user is non-deterministicaly asked to select a branch when stepping forwards over the `JUMPI` on PC13 or PC18.

### PR DESCRIPTION
**Pull request by MrChico**
Fixes #456. I previously failed to add the knowledge gained from queries returning `Known`, thinking that if this only happens when there is only one alternative, in which case we are not really adding any new information. But `Known` can also be the result of a cached choice, so the newly gained knowledge should be stored.

### CODE REVIEW

**Comment by MrChico** [comment]
hmm, but this actually screws things up

**Comment by MrChico** [comment]
Okay, I found the cause of the soundness issue, it's a one character change (65138d1). This is included in this PR and I also removed the test case that found this from the list of ignores

**Comment by MrChico** [comment]
Still not quite right...

**Comment by MrChico** [comment]
Okay, all the failures are only timeouts now

**Comment by MrChico** [comment]
ah, but now it's possible to end up in an inifite loop

**Comment by MrChico** [comment]
Okay, I'm satisfied with this now. Please review if you can @xwvvvvwx. I had to ignore some new tests because things can become a little slower, because I'm taking some extra precautions now when branching:
At each JUMPI, after checking if `jumpcondition /= 0` is satisfiable, I now always also check its negation `jumpcondition == 0`, and end the path with failure if neither condition is satisfiable. This happens if we end up in a path with inconsistent path conditions, which can happen for example because of a previous timeout.

In summary, this PR changes the following things:
- When the jumpcondition is nonzero, assume that it is nonzero instead of assuming that it is one.
- Always add the knowledge gained from taking a path. Fixes #456
- End paths with failure if their pathconditions are inconsistent.
- Fix the embedding of bytestrings into integers, resolving the injectivity problem with keccak that causes #466

File: src/hevm/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,8 @@
 ### Changed
 
 - z3 updated to 4.8.7
+- Fixed soundness bug when using KECCAK or SHA256 opcode/precompile
+- Fixed an issue in debug mode where backstepping could cause path information to be forgotten
```
Comment (MrChico) [inline src/hevm/CHANGELOG.md:10]:
```suggestion
- Fixed an issue in debug mode where backstepping could cause path information to be forgotten
- Ensure that pathconditions are consistent when branching, and end the execution with VMFailure: DeadPath if this is not the case
```

File: src/hevm/CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,9 @@
...
+- Ensure that pathconditions are consistent when branching, and end the execution with VMFailure: DeadPath if this is not the case
```
Comment (MrChico) [inline src/hevm/CHANGELOG.md:11]:
```suggestion
- Ensure that pathconditions are consistent when branching, and end the execution with VMFailure: DeadPath if this is not the case
- Fixed a soundness bug where nonzero jumpconditions  were assumed to equal one.
```

File: src/hevm/src/EVM/Fetch.hs [hunk 3]
```diff
@@ -167,23 +167,22 @@ oracle state info q = do
...
+                          -- Assume 0 is still possible.
+                          Unk -> return $ continue (EVM.Iszero True)
             -- Sat means its possible for condition
             -- to be nonzero.
             Sat -> do jump <- checksat $ pathconds .&& jumpcondition .== 0
                       -- can it also be zero?
                       case jump of
                         -- No. It must be nonzero
-                        Unsat -> return $ continue (EVM.Known 1)
+                        Unsat -> return $ continue (EVM.Iszero False)
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM/Fetch.hs:185]: If I understood you correctly, this was another soundness bug @MrChico?

Could we add a test case that covers this behaviour?

Comment (MrChico) [reply src/hevm/src/EVM/Fetch.hs:185]: I discovered it through [this test](https://github.com/ethereum/solidity/blob/develop/test/libsolidity/smtCheckerTests/operators/compound_assignment_division_1.sol)

Comment (MrChico) [reply src/hevm/src/EVM/Fetch.hs:185]: Which was previously not a problem because we weren't storing the path condition information upon branching when the query was Known, but now we are adding this information anyway so that test should be sufficient 👍
